src/systems/gimbal_controller/GimbalControllerPlugin.cc

```diff
@@ -0,0 +1,490 @@
+}
+
+////////////////////////////////////////////////////////////////////////////
+ignition::math::Vector3d GimbalControllerPluginPrivate::ThreeAxisRot(double r11,
+   double r12, double r21, double r31, double r32)
+{
+   return ignition::math::Vector3d(atan2
+      (r31, r32), asin (r21), atan2(r11, r12));
+}
+
+////////////////////////////////////////////////////////////////////////////
+ignition::math::Vector3d GimbalControllerPluginPrivate::QtoZXY(
+   const ignition::math::Quaterniond &_q)
+{
+   // taken from
+   // http://bediyap.com/programming/convert-quaternion-to-euler-rotations/
+   // case zxy:
+   ignition::math::Vector3d result = this->ThreeAxisRot(
+     -2*(_q.X()*_q.Y() - _q.W()*_q.Z()),
+     _q.W()*_q.W() - _q.X()*_q.X() + _q.Y()*_q.Y() - _q.Z()*_q.Z(),
+     2*(_q.Y()*_q.Z() + _q.W()*_q.X()),
+     -2*(_q.X()*_q.Z() - _q.W()*_q.Y()),
+     _q.W()*_q.W() - _q.X()*_q.X() - _q.Y()*_q.Y() + _q.Z()*_q.Z());
+   return result;
+}
+
+////////////////////////////////////////////////////////////////////////////
+void GimbalControllerPlugin::PreUpdate(const UpdateInfo &_info,
+    ignition::gazebo::EntityComponentManager &_ecm)
+{
+   if(!this->dataPtr->imuInitialized)
+   {
+      // Set unconditionally because we're only going to try this once.
+      this->dataPtr->imuInitialized = true;
+
+      auto imuEntity = _ecm.EntityByComponents(
+          components::Name(this->dataPtr->imuName),
+          components::Imu(),
+          components::ParentEntity(this->dataPtr->modelLink));
+
+      // Imu sensor topic
+      std::string topic = scopedName(imuEntity, _ecm) + "/imu";
+
+      if(topic.empty())
+      {
+         ignerr << "imu_sensor [" << this->dataPtr->imuName
+                << "] not found, abort ArduPilot plugin." << "\n";
+         return;
+      }
+      this->dataPtr->node.Subscribe(topic,
+         &GimbalControllerPluginPrivate::imuCb);
+   }
+
+   if(!this->dataPtr->pitchJoint || !this->dataPtr->rollJoint ||
+      !this->dataPtr->yawJoint)
+      return;
+
+   if(_info.simTime < this->dataPtr->lastControllerUpdateTime)
+   {
+      ignerr << "time reset event\n";
+      this->dataPtr->lastControllerUpdateTime = _info.simTime;
+      return;
+   }
+   else if(_info.simTime > this->dataPtr->lastControllerUpdateTime)
+   {
+      // Time delta
+      std::chrono::duration<double> dt =
+      (this->dataPtr->lastControllerUpdateTime - _info.simTime);
+
+      // pitch joint position
+      auto pitchJointPos = _ecm.Component<components::JointPosition>(
+         this->dataPtr->pitchJoint)->Data().at{0};
+
+      ignition::gazebo::components::JointForceCmd* pjfcComp = nullptr;
+      pjfcComp = _ecm.Component<components::JointForceCmd>(
+         this->dataPtr->pitchJoint);
+
+      if (pjfcComp == nullptr)
+      {
+         pjfcComp = _ecm.Component<components::JointForceCmd>(
+            _ecm.CreateComponent(this->dataPtr->pitchJoint,
+               components::JointForceCmd({0})));
```

Comment:
I think this should looks like

```suggestion
        *pjfcComp = ignition::gazebo::components::JointForceCmd(
          <joint_force_cmd>);
```